identity/src/secp256k1.rs

```diff
@@ -166,58 +166,55 @@ impl fmt::Debug for PublicKey {
     }
 }
 
-impl cmp::PartialEq for PublicKey {
-    fn eq(&self, other: &Self) -> bool {
-        self.to_bytes().eq(&other.to_bytes())
-    }
-}
-
 impl hash::Hash for PublicKey {
     fn hash<H: hash::Hasher>(&self, state: &mut H) {
         self.to_bytes().hash(state);
     }
 }
 
-impl cmp::PartialOrd for PublicKey {
-    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
-        Some(self.cmp(other))
-    }
-}
-
-impl cmp::Ord for PublicKey {
-    fn cmp(&self, other: &Self) -> cmp::Ordering {
-        self.to_bytes().cmp(&other.to_bytes())
-    }
-}
-
 impl PublicKey {
     /// Verify the Secp256k1 signature on a message using the public key.
     pub fn verify(&self, msg: &[u8], sig: &[u8]) -> bool {
-        self.verify_hash(Sha256::digest(msg).as_ref(), sig)
+        let digest = Sha256::new_with_prefix(msg);
+        self.verify_hash(digest.finalize().as_slice(), sig)
     }
 
-    /// Verify the Secp256k1 DER-encoded signature on a raw 256-bit message using the public key.
+    /// Verify the Secp256k1 DER-encoded signature on a raw 256-bit message using the public key.  
+    /// Will panic if the hash is not 32 bytes long.
     pub fn verify_hash(&self, msg: &[u8], sig: &[u8]) -> bool {
-        Message::parse_slice(msg)
-            .and_then(|m| Signature::parse_der(sig).map(|s| libsecp256k1::verify(&m, &s, &self.0)))
-            .unwrap_or(false)
+        Signature::from_der(sig).is_ok_and(|s| {
+            k256::ecdsa::hazmat::verify_prehashed(
+                &ProjectivePoint::from(self.0.as_affine()),
+                GenericArray::from_slice(msg),
+                &s,
+            )
+            .is_ok()
+        })
     }
 
     /// Convert the public key to a byte buffer in compressed form, i.e. with one coordinate
     /// represented by a single bit.
-    pub fn to_bytes(&self) -> [u8; 33] {
-        self.0.serialize_compressed()
+    pub fn to_bytes(self) -> [u8; 33] {
+        let encoded_point = self.0.to_encoded_point(true);
+        debug_assert!(encoded_point.as_bytes().len() == 33);
+        let mut array: [u8; 33] = [0u8; 33];
+        array.copy_from_slice(encoded_point.as_bytes());
+        array
     }
 
     /// Convert the public key to a byte buffer in uncompressed form.
-    pub fn to_bytes_uncompressed(&self) -> [u8; 65] {
-        self.0.serialize()
+    pub fn to_bytes_uncompressed(self) -> [u8; 65] {
+        let encoded_point = self.0.to_encoded_point(false);
+        debug_assert!(encoded_point.as_bytes().len() == 65);
+        let mut array: [u8; 65] = [0u8; 65];
+        array.copy_from_slice(encoded_point.as_bytes());
+        array
     }
```

Comment:
These are breaking changes. Can we keep taking a reference to `self` rather than ownership?

Reply:
Breaking changes in `identity` are quite painful because it bubbles up through every single crate in `libp2p` that then each need a minor version bump and release. We are really trying to avoid that. 

Thinking about it now, I think we shouldn't even add the `Copy` derive because it could cause a lot of warnings in downstream code if we don't release it as a breaking change.

Reply:
cc @jxs since you reacted

Reply:
thanks for the ping DrHuang, yeah agree with Elena, no need to add the `Copy` derive the `Clone`, in this case it will only be sugar as `Clone` will just `*self` right?